### PR TITLE
Resolve and unpack operators from bundle images

### PR DESCRIPTION
**Pull request by jpeeler**

### CODE REVIEW

File: pkg/controller/registry/resolver/resolver.go [hunk 1]
```diff
@@ -23,18 +24,20 @@ type Resolver interface {
...
+	subLister  v1alpha1listers.SubscriptionLister
+	csvLister  v1alpha1listers.ClusterServiceVersionLister
+	client     versioned.Interface
+	kubeclient kubernetes.Interface
 }
 
 var _ Resolver = &OperatorsV1alpha1Resolver{}
 
-func NewOperatorsV1alpha1Resolver(lister operatorlister.OperatorLister, client versioned.Interface) *OperatorsV1alpha1Resolver {
+func NewOperatorsV1alpha1Resolver(lister operatorlister.OperatorLister, client versioned.Interface, kubeclient kubernetes.Interface) *OperatorsV1alpha1Resolver {
```
Comment (njhale) [inline pkg/controller/registry/resolver/resolver.go]: If we think the parameter list will keep growing, then it's probably worth it to use the variadic config options pattern here.

On the flip side, we could also figure out how to compose typed clientsets nicely (see https://github.com/kubernetes/client-go/blob/0bdba2f9188006fc64057c2f6d82a0f9ee0ee422/kubernetes/scheme/register.go#L121).

File: pkg/controller/registry/resolver/querier.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ type SourceRef struct {
 
 type SourceQuerier interface {
 	FindProvider(api opregistry.APIKey) (*opregistry.Bundle, *CatalogKey, error)
-	FindBundle(pkgName, channelName, bundleName string, initialSource CatalogKey) (*opregistry.Bundle, *CatalogKey, error)
+	FindBundle(pkgName, channelName, bundleName string, initialSource CatalogKey) (*opregistry.Bundle, *CatalogKey, *string, error)
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/controller/registry/resolver/querier.go]: This is implemented in the the pr that just merged: https://github.com/operator-framework/operator-registry/pull/103

File: pkg/controller/registry/resolver/evolver.go [hunk 3]
```diff
@@ -78,9 +78,25 @@ func (e *NamespaceGenerationEvolver) addNewOperators(add map[OperatorSourceInfo]
...
 		var key *CatalogKey
+		var bundlePath *string
 		var err error
 		if s.StartingCSV != "" {
-			bundle, key, err = e.querier.FindBundle(s.Package, s.Channel, s.StartingCSV, s.Catalog)
+			bundle, key, bundlePath, err = e.querier.FindBundle(s.Package, s.Channel, s.StartingCSV, s.Catalog)
+			if bundlePath != nil {
+				// TODO: downward API for namespace
+				// TODO: kubernetes client
+				configMap, job, err := LaunchBundleImage(o.client, bundlePath, "quay.io/rht/init-bundle-image:latest", "olm")
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/evolver.go]: This is obsolete.

File: pkg/controller/registry/resolver/generation.go [hunk 4]
```diff
@@ -67,6 +81,10 @@ func NewGenerationFromCluster(csvs []*v1alpha1.ClusterServiceVersion, subs []*v1
 	return g, nil
 }
 
+func (g *NamespaceGeneration) AddPendingOperator(l LaunchBundleImageInfo) {
+	g.pendingOperators = append(g.pendingOperators, l)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/generation.go]: Yep, already fixed.

File: pkg/controller/registry/resolver/evolver.go [hunk 5]
```diff
@@ -78,9 +78,25 @@ func (e *NamespaceGenerationEvolver) addNewOperators(add map[OperatorSourceInfo]
 	for s := range add {
 		var bundle *opregistry.Bundle
 		var key *CatalogKey
+		var bundlePath *string
 		var err error
 		if s.StartingCSV != "" {
-			bundle, key, err = e.querier.FindBundle(s.Package, s.Channel, s.StartingCSV, s.Catalog)
+			bundle, key, bundlePath, err = e.querier.FindBundle(s.Package, s.Channel, s.StartingCSV, s.Catalog)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/evolver.go]: Well, bundlePath is the same name as definied in operator-registry. Do you still think the name should be different?

File: pkg/controller/registry/resolver/generation.go [hunk 6]
```diff
@@ -18,13 +22,23 @@ type Generation interface {
...
 }
 
+type LaunchBundleImageInfo struct {
+	operatorSourceInfo OperatorSourceInfo
+	job                *batchv1.Job
+	configMap          *corev1.ConfigMap
+	image              *string
+}
+
+type BundleImageSet map[LaunchBundleImageInfo]struct{}
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/generation.go]: Possibly... will circle back.

File: pkg/controller/registry/resolver/querier.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/querier.go]: Yeah this is obsolete.

File: pkg/controller/registry/resolver/generation.go [hunk 7]
```diff
@@ -3,13 +3,17 @@ package resolver
...
+	corev1 "k8s.io/api/core/v1"
+
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
 	"github.com/operator-framework/operator-registry/pkg/registry"
 )
 
 // Generation represents a set of operators and their required/provided API surfaces at a point in time.
 type Generation interface {
 	AddOperator(o OperatorSurface) error
+	AddPendingOperator(l LaunchBundleImageInfo)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/generation.go]: Will circle back to this.

File: pkg/controller/registry/resolver/generation.go [hunk 8]
```diff
@@ -3,13 +3,17 @@ package resolver
 import (
 	"fmt"
 
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
 	"github.com/operator-framework/operator-registry/pkg/registry"
```
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/controller/registry/resolver/generation.go]: We've standardized on `goimports` -- there should be an option in your `vscode-go` config to use it.

File: pkg/controller/registry/resolver/evolver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/controller/registry/resolver/evolver.go]: I think it would be more clear to change it in both places if I'm understanding its purpose correctly.

File: pkg/controller/operators/catalog/operator.go [hunk 9]
```diff
@@ -1014,6 +1021,71 @@ func (o *Operator) createInstallPlan(namespace string, subs []*v1alpha1.Subscrip
...
+		// combine data from the bundle image into what's already known from the registry
+		bundleLookup.BundleFromRegistry.CsvName = manifest.Bundle.Name
+		bundleLookup.BundleFromRegistry.PackageName = manifest.Bundle.Package
+		bundleLookup.BundleFromRegistry.ChannelName = manifest.Bundle.Channel
+		_, jsonCSV, _, err := manifest.Bundle.Serialize()
+		if err != nil {
+			return false, fmt.Errorf("serialize failed: %s", err.Error())
+		}
+		bundleLookup.BundleFromRegistry.CsvJson = string(jsonCSV)
+		bundleLookup.BundleFromRegistry.Object, err = registry.BundleStringToObjectStrings(bundleLookup.BundleFromRegistry.CsvJson)
```
Comment (ecordell) [inline pkg/controller/operators/catalog/operator.go]: I think this can just be `bundleLookup.BundleFromRegistry.Object = []string{string(jsonCSV)}`

File: pkg/controller/operators/catalog/operator.go [hunk 10]
```diff
@@ -132,12 +137,14 @@ func NewOperator(ctx context.Context, kubeconfigPath string, clock utilclock.Clo
 		client:                 crClient,
 		lister:                 lister,
 		namespace:              operatorNamespace,
-		resolver:               resolver.NewOperatorsV1alpha1Resolver(lister, crClient),
+		resolver:               resolver.NewOperatorsV1alpha1Resolver(lister, crClient, opClient.KubernetesInterface(), operatorNamespace),
```
Comment (ecordell) [inline pkg/controller/operators/catalog/operator.go]: note to self: why is this needed?

File: pkg/controller/operators/catalog/operator.go (hunk 9 above)
Comment (ecordell) [reply pkg/controller/operators/catalog/operator.go]: ^ still this 😄

File: pkg/controller/registry/resolver/resolver.go [hunk 11]
```diff
@@ -129,7 +140,57 @@ func (r *OperatorsV1alpha1Resolver) ResolveSteps(namespace string, sourceQuerier
...
+
+	// prune operators that have in progress bundle image jobs
+	for bundleImageInfo := range gen.PendingOperators() {
+		// TODO: this is not ideal...
+		ips, err := r.listInstallPlans(namespace)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for _, ip := range ips {
+			for _, lookup := range ip.Status.BundleLookups {
```
Comment (ecordell) [inline pkg/controller/registry/resolver/resolver.go]: I think I'm missing why this is needed - can't you just return the whole set of pending operators as `BundleLookups` that the installplan should unpack?

File: pkg/controller/registry/resolver/resolver.go [hunk 12]
```diff
@@ -129,7 +140,57 @@ func (r *OperatorsV1alpha1Resolver) ResolveSteps(namespace string, sourceQuerier
...
+					gen.RemovePendingOperator(bundleImageInfo)
+				}
+			}
+		}
+	}
+
+	for bundleImageInfo := range gen.PendingOperators() {
+		// TODO: switch image to standalone image, but this image can be used upstream as well
+		// change to use configmapRegistryImage
+		//configmap, job, err := configmap.LaunchBundleImage(r.kubeclient, bundleImageInfo.image, "quay.io/openshift/origin-operator-registry:latest", r.operatorNamespace)
```
Comment (ecordell) [inline pkg/controller/registry/resolver/resolver.go]: to clarify our discussion last night, you should use the actual variable "configmapRegistryImage" from `pkg/catalogoperator.go NewOperator`, not `quay.io/openshift/origin-operator-registry:latest`

File: pkg/lib/queueinformer/resourcequeue.go [hunk 13]
```diff
@@ -68,6 +69,29 @@ func (r *ResourceQueueSet) Requeue(namespace, name string) error {
 	return fmt.Errorf("couldn't find queue for resource")
 }
 
+// TODO: this may not actually be required if the requeue is done on the namespace rather than the installplan
+// RequeueAfter requeues the resource in the set with the given name and namespace (just like Requeue), but only does so after duration has passed
+func (r *ResourceQueueSet) RequeueAfter(namespace, name string, duration time.Duration) error {
```
Comment (ecordell) [inline pkg/lib/queueinformer/resourcequeue.go]: Can we watch `Jobs` and requeue owner installplans when we see them?

File: pkg/controller/registry/resolver/resolver.go [hunk 14]
```diff
@@ -129,7 +140,57 @@ func (r *OperatorsV1alpha1Resolver) ResolveSteps(namespace string, sourceQuerier
...
+
+		configmap, job, err := configmap.LaunchBundleImage(r.kubeclient, bundleImageInfo.image, "quay.io/jpeeler/bundle-init-image:latest", r.operatorNamespace)
```
Comment (ecordell) [inline pkg/controller/registry/resolver/resolver.go]: where do these get cleaned up?

(this is not necessarily blocking, we can follow up with GC)

Comment (ecordell) [reply pkg/controller/registry/resolver/resolver.go]: we could return the pending operators from the resolver without launching them?

then when we create the installplan, write the "bundlelookups" into it

then when we process the installplan, launch jobs for any bundlelookups that don't have corresponding jobs?

this would also mean that when we launch, we have an installplan, so we can add ownerrefs, so we can watch jobs and requeue based on ownerref 😄

Comment (jpeeler) [reply pkg/controller/registry/resolver/resolver.go]: you are correct that they aren't yet

File: pkg/controller/operators/catalog/operator.go (hunk 10 above)
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: It's used for launching the bundle image (LaunchBundleImage).

File: pkg/controller/operators/catalog/operator.go (hunk 9 above)
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: Yeah good call. Anything to reduce potential errors here.

File: pkg/controller/registry/resolver/resolver.go (hunk 11 above)
Comment (jpeeler) [reply pkg/controller/registry/resolver/resolver.go]: I do have a TODO here to investigate improving. Without it install plans get duplicated due to the next loop doing a resolve (again) and repeating. The only way that I can see to avoid this is by checking all the install plans for the image.

File: pkg/lib/queueinformer/resourcequeue.go (hunk 13 above)
Comment (jpeeler) [reply pkg/lib/queueinformer/resourcequeue.go]: Yeah it's a good idea and is something I was considering.

**Comment by jpeeler** [comment]
This isn't merge-able :weary:

The e2e currently has exactly double steps, which seems like resources are getting added from the bundle (which in this case has a CSV) and the bundle image.

Code generation has not been done, unit tests aren't updated.

**Comment by jpeeler** [comment]
/hold
This isn't quite ready yet, but is much closer now. I tried to properly integrate the ideas from all of these PRs here as well:

https://github.com/operator-framework/operator-lifecycle-manager/pull/1194
https://github.com/operator-framework/operator-lifecycle-manager/pull/1105
https://github.com/operator-framework/operator-lifecycle-manager/pull/1108

This is a blocking PR that needs to be merged first (but could possibly change due to feedback received here):
https://github.com/operator-framework/operator-registry/pull/137

File: pkg/controller/operators/catalog/operator.go [hunk 15]
```diff
@@ -85,6 +87,7 @@ type Operator struct {
 	catalogSubscriberIndexer map[string]cache.Indexer
 	clientAttenuator         *scoped.ClientAttenuator
 	serviceAccountQuerier    *scoped.UserDefinedServiceAccountQuerier
+	bundleLoader             *configmap.BundleLoader
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: I don't know, Abu actually wrote the v1 BundleLoader in operator-registry. I assume it was just modeled after the sqlite loader.

File: pkg/controller/operators/catalog/operator.go [hunk 16]
```diff
@@ -769,7 +777,7 @@ func (o *Operator) syncResolvingNamespace(obj interface{}) error {
 	logger.Debug("resolving subscriptions in namespace")
 
 	// resolve a set of steps to apply to a cluster, a set of subscriptions to create/update, and any errors
-	steps, updatedSubs, err := o.resolver.ResolveSteps(namespace, querier)
+	steps, bundleLookups, updatedSubs, err := o.resolver.ResolveSteps(namespace, querier)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: The reason I did this was specifically to avoid modifying the flow for install plan creation as little as possible. But maybe it's just too ugly.

File: pkg/controller/operators/catalog/operator.go [hunk 17]
```diff
@@ -1033,6 +1042,84 @@ func (o *Operator) createInstallPlan(namespace string, subs []*v1alpha1.Subscrip
 	return reference.GetReference(res)
 }
 
+func (o *Operator) checkBundleLookups(plan *v1alpha1.InstallPlan) (bool, error) {
+	for _, bundleLookup := range plan.Status.BundleLookups {
+		if bundleLookup.BundleJob == nil {
+			configmap, job, err := configmap.LaunchBundleImage(o.opClient.KubernetesInterface(), bundleLookup.Image, "quay.io/jpeeler/bundle-init-image:latest", o.namespace)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: Do you mean it should remain its own image entirely? A while back I added the opm binary to the openshift image: https://github.com/operator-framework/operator-registry/commit/9d0036ae2b81ae8e531212dca88cb9feeaf128c2. For upstream I was considering adding it to the configmap-operator-registry image. 

But definitely this image needs to be changed to something official and maintained.

File: pkg/controller/operators/catalog/operator.go [hunk 18]
```diff
@@ -1033,6 +1042,84 @@ func (o *Operator) createInstallPlan(namespace string, subs []*v1alpha1.Subscrip
 	return reference.GetReference(res)
 }
 
+func (o *Operator) checkBundleLookups(plan *v1alpha1.InstallPlan) (bool, error) {
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: Once a lookup has been completed, the completion time is updated in the status to avoid future work.

File: pkg/controller/operators/catalog/operator.go [hunk 19]
```diff
@@ -1033,6 +1042,84 @@ func (o *Operator) createInstallPlan(namespace string, subs []*v1alpha1.Subscrip
...
+				Name:      job.GetName(),
+				Namespace: job.GetNamespace(),
+			}
+			bundleLookup.ConfigMapRef = &v1alpha1.ConfigMapResourceReference{
+				Name:            configmap.GetName(),
+				Namespace:       configmap.GetNamespace(),
+				UID:             configmap.GetUID(),
+				ResourceVersion: configmap.GetResourceVersion(),
+			}
+			_, err = o.client.OperatorsV1alpha1().InstallPlans(plan.GetNamespace()).UpdateStatus(plan)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: The idea was to reduce the possibility of having untracked launched jobs. If a job launch fails, but previous ones have succeeded, then additional logic has to be done to make sure the successful ones get written before erroring.

File: pkg/controller/operators/catalog/operator.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/operators/catalog/operator.go]: Added here: https://github.com/operator-framework/operator-registry/pull/142

File: pkg/controller/registry/resolver/resolver.go [hunk 20]
```diff
@@ -185,3 +200,17 @@ func (r *OperatorsV1alpha1Resolver) listSubscriptions(namespace string) (subs []
...
 	return
 }
+
+func (r *OperatorsV1alpha1Resolver) listInstallPlans(namespace string) (ips []*v1alpha1.InstallPlan, err error) {
+	list, err := r.client.OperatorsV1alpha1().InstallPlans(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return
+	}
+
+	ips = make([]*v1alpha1.InstallPlan, 0)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/resolver.go]: How about I delete the entire function?

File: test/e2e/installplan_e2e_test.go [hunk 21]
```diff
@@ -2549,3 +2566,123 @@ func TestInstallPlanCRDValidation(t *testing.T) {
...
+				Namespace: testNamespace,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     rbacRole.GetName(),
+		},
+	}
+	_, err = c.KubernetesInterface().RbacV1().RoleBindings(testNamespace).Create(rbacBinding)
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply test/e2e/installplan_e2e_test.go]: This might actually only be necessary for local dev testing. In a production cluster, OLM should have writes to read configmaps in any namespace, right?

File: pkg/controller/registry/resolver/resolver.go [hunk 22]
```diff
@@ -93,10 +99,10 @@ func (r *OperatorsV1alpha1Resolver) ResolveSteps(namespace string, sourceQuerier
 		}
 
 		// add steps for any new bundle
-		if op.Bundle() != nil {
+		if op.Bundle() != nil && op.Bundle().BundlePath == "" {
```
[the remark this answers is not in the document]
Comment (jpeeler) [reply pkg/controller/registry/resolver/resolver.go]: NewOperatorFromBundle assumes that the returned bundle isn't nil shortly after retrieving the bundle in, so I think checking for this is unnecessary anyway.

File: pkg/controller/operators/catalog/operator.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/controller/operators/catalog/operator.go]: That's a fair reason. I'd be happy w/ any solution that avoids parameter spaghettification across multiple methods/functions.

File: pkg/controller/operators/catalog/operator.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/controller/operators/catalog/operator.go]: This leads me to think that the problem is that we can have untracked launched jobs. I think we should be able to create jobs in such a way that we discover them later w/o relying on a separate resource (i.e. `InstallPlan`).

File: pkg/controller/operators/catalog/operator.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/controller/operators/catalog/operator.go]: I primarily meant that, code-wise, it would be nice for the image to be an optional configuration.

The other part of things is that the default should at least be under the `operator-framework` Quay org, which it looks like you've addressed already.

File: pkg/controller/operators/catalog/operator.go [hunk 23]
```diff
@@ -1033,6 +1044,96 @@ func (o *Operator) createInstallPlan(namespace string, subs []*v1alpha1.Subscrip
 	return reference.GetReference(res)
 }
 
+func (o *Operator) checkBundleLookups(plan *v1alpha1.InstallPlan) (bool, error) {
+	for _, bundleLookup := range plan.Status.BundleLookups {
```
Comment (njhale) [inline pkg/controller/operators/catalog/operator.go]: Over the course of our work on OLM, we've developed an idiomatic way of reconciling resources:

1. Build state from cluster
2. Inform cluster action from state; or
3. Update resource status from state

I think that we can run into trouble if we jump straight to 2. without first performing 1., or by coupling 2. and 3.

In this case, do you think we can:

- build the state before actioning on it; i.e. Check for `ConfigMaps`/`Jobs` before creating them
- update the `InstallPlan` object once per reconciliation

File: pkg/controller/operators/catalog/operator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/controller/operators/catalog/operator.go]: I think in an ideal world this would be an interface. It would be nice if we had other options than unpacking into a configmap in the future. But we can do that as a separate refactoring after this merges.

File: pkg/controller/operators/catalog/operator.go [hunk 24]
```diff
@@ -133,13 +137,16 @@ func NewOperator(ctx context.Context, kubeconfigPath string, clock utilclock.Clo
...
+		resolver:                 resolver.NewOperatorsV1alpha1Resolver(lister, crClient, opClient.KubernetesInterface()),
 		catsrcQueueSet:           queueinformer.NewEmptyResourceQueueSet(),
 		subQueueSet:              queueinformer.NewEmptyResourceQueueSet(),
+		ipQueueSet:               queueinformer.NewEmptyResourceQueueSet(),
 		csvProvidedAPIsIndexer:   map[string]cache.Indexer{},
 		catalogSubscriberIndexer: map[string]cache.Indexer{},
 		serviceAccountQuerier:    scoped.NewUserDefinedServiceAccountQuerier(logger, crClient),
 		clientAttenuator:         scoped.NewClientAttenuator(logger, config, opClient, crClient),
+		bundleLoader:             configmap.NewBundleLoader(),
+		configmapRegistryImage:   configmapRegistryImage,
```
Comment (ecordell) [inline pkg/controller/operators/catalog/operator.go]: We probably want a separate variable for "bundleUnpackerImage". In OpenShift this will end up the same image (registry image) but they do different things and there's no reason they should have to be the same.

File: pkg/controller/operators/catalog/operator.go (hunk 23 above)
Comment (ecordell) [reply pkg/controller/operators/catalog/operator.go]: This is a good plan - we just need a deterministic way to name the configmaps/jobs from the installplan. InstallPlanName + image name would make sense, or a hash of that info.

File: pkg/controller/operators/catalog/operator.go [hunk 25]
```diff
@@ -1033,6 +1042,84 @@ func (o *Operator) createInstallPlan(namespace string, subs []*v1alpha1.Subscrip
...
+		bundleLookup.BundleJob.Condition = job.Status.Conditions[0].Type
+		bundleLookup.BundleJob.CompletionTime = job.Status.CompletionTime
+
+		configmap, err := o.opClient.KubernetesInterface().CoreV1().ConfigMaps(bundleLookup.ConfigMapRef.Namespace).Get(bundleLookup.ConfigMapRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		// extract data from configmap and write to install plan
+		bundle, err := o.bundleLoader.Load(configmap)
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/controller/operators/catalog/operator.go]: That makes sense but is not critical, IMO. Installing via installplan steps ensures we install via a well-tested path. But, this would be a great followup and will enable installplans to install larger sets of objects.

**Comment by jpeeler** [comment]
Superseded by #1215, which has now merged.
